proparse/src/main/java/org/prorefactor/proparse/antlr4/PreproEval.java

```diff
@@ -0,0 +1,612 @@
+        throw new RuntimeException();
+    }
+  }
+
+  /**
+   * Use compareTo() from String, Integer, or Float for two objects. Returns 0 for comparison of two nulls. Returns null
+   * if only one of the two is null.
+   */
+  private static Integer compare(Object left, Object right) {
+    if (left == null && right == null)
+      return 0;
+    if (left == null || right == null)
+      return null;
+    if ((left instanceof Boolean) && (right instanceof Boolean))
+      return ((Boolean) left).compareTo((Boolean) right);
+    if ((left instanceof String) && (right instanceof String))
+      return compareStringHelper(left).compareTo(compareStringHelper(right));
+    if ((left instanceof Number) && (right instanceof Number)) {
+      Double fl = ((Number) left).doubleValue();
+      Double fr = ((Number) right).doubleValue();
+      return fl.compareTo(fr);
+    }
+    throw new ProEvalException("Incompatible data types in comparison expression.");
+  }
+
+  static String compareStringHelper(Object o) {
+    // Lowercase and right-trim the string for comparison, like Progress does.
+    return StringFuncs.rtrim(((String) o).toLowerCase());
+  }
+
+  static Float decimal(Object o) {
+    if (o == null)
+      return null;
+    if (o instanceof Number)
+      return ((Number) o).floatValue();
+    if (o instanceof String)
+      return getNumber((String) o).floatValue();
+    if (o instanceof Boolean)
+      return (Boolean) o ? 1f : 0f;
+    throw new ProEvalException("Error converting to DECIMAL.");
+  }
+
+  static String entry(Object oa, Object ob, Object odelim) {
+    if (oa == null || ob == null)
+      return null;
+    int pos = getInt(oa);
+    String b = getString(ob);
+    String delim;
+    if (odelim == null)
+      delim = ",";
+    else
+      delim = getString(odelim);
+    // Progress position numbers start at 1
+    if (pos < 1)
+      throw new ProEvalException("ENTRY function received non-positive number");
+    delim = delim.substring(0, 1);
+    Pattern regex = Pattern.compile(delim, Pattern.LITERAL | Pattern.CASE_INSENSITIVE);
+    String[] array = regex.split(b);
+    if (pos > array.length)
+      return "";
+    return array[pos - 1];
+  }
+
+  static boolean getBool(Object obj) {
+    // Implicit conversion from int or ProString to bool
+    // Note that Progress does /not/ do implicit conversion to bool from decimal.
+    if (obj instanceof String)
+      return ((String) obj).length() != 0;
+    if (obj instanceof Boolean)
+      return (Boolean) obj;
+    if (obj instanceof Integer)
+      return ((Integer) obj) != 0;
+    throw new ProEvalException("Unknown datatype passed to getBool");
+  }
+
+  static float getFloat(Object obj) {
+    // Implicit conversion from int to float, but no others.
+    if (obj instanceof Float)
+      return (Float) obj;
+    if (obj instanceof Integer)
+      return (Integer) obj;
+    throw new ProEvalException("Incompatible datatype");
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Define a constant instead of duplicating this literal "Incompatible datatype" 3 times. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://sonar.riverside-software.fr/coding_rules#rule_key=squid%3AS1192)